src/types.lisp

```diff
@@ -124,6 +124,60 @@ If NAME is not known, it will be made known to the global type database."
 
 #+sbcl (declaim (sb-ext:freeze-type ty tyvar tyapp tyfun))
 
+(defun type= (type1 type2)
+  "Check equality of types TYPE1 and TYPE2
+
+Types are equivalent when the structure (TYAPP and TYFUN) matches and there exists a bijection between TYVARs of each type."
+  (declare (type ty type1 type2)
+           (values boolean))
+  (let ((var-table nil))
+    (labels ((%type= (ty1 ty2)
+               (let ((pty1 (prune ty1))
+                     (pty2 (prune ty2)))
+                 (cond
+                   ((and (tyvar-p pty1)
+                         (tyvar-p pty2))
+                    (let ((pair1
+                            (find-if
+                             (lambda (x)
+                               (= (tyvar-id pty1)
+                                  (car x)))
+                             var-table))
+                          (pair2
+                            (find-if
+                             (lambda (x)
+                               (= (tyvar-id pty2)
+                                  (cadr x)))
+                             var-table)))
+                      (cond
+                        ((equal pair1 pair2)
+                         (when (null pair1)
+                           (push (list (tyvar-id pty1) (tyvar-id pty2))
+                                 var-table))
+                         t)
+                        (t
+                         nil))))
+                   ((and (tyfun-p pty1)
+                         (tyfun-p pty2))
+                    (let ((arity-1 (length (tyfun-from pty1)))
```

Comment:
`tyfun-arity` is there for ur taking

Reply:
(which means you can collapse the LET and change the last bit to
```
(and (= arity ...) (type= ...) (every ...))
```
without the `if`